### ISSUE TITLE
bug: eval dataset not being handled correctly in latest version

### ISSUE THREAD

**Issue by Alex-Shan** [PrimeIntellect-ai/prime-rl#2871]
<hr>
<h2>Description</h2>
<h3>Problem</h3>
<p>When an environment is constructed with both a <code>dataset</code> and an <code>eval_dataset</code>, the orchestrator's eval environment (<code>EvalEnv</code>) still reports <code>num_tasks</code> from the <strong>training</strong> dataset rather than the dedicated eval split.</p>
<h3>Reproduction</h3>
<p><strong><code>train.toml</code></strong></p>
<pre><code class="language-toml">[[orchestrator.train.env]]
id = "or_v4_5"
max_retries = 10
sampling = {temperature=1.0, extra_body={top_k=20, min_p=0.0, presence_penalty=1.5, repetition_penalty=1.0}}

[[orchestrator.eval.env]]
id = "or_v4_5"
interval = 50
sampling = {temperature=0.6, top_p=0.95, top_k=20, min_p=0.0, extra_body={presence_penalty=1.5, repetition_penalty=1.0}}
</code></pre>
<p><strong>Environment constructor (inside <code>or_v4_5</code>)</strong></p>
<pre><code class="language-python">return SandboxEnv(
    dataset=train_dataset,      # 92 examples
    eval_dataset=test_dataset,  # 100 examples
    ...
)
</code></pre>
<p><strong>Observed orchestrator log</strong></p>
<pre><code>03:37:53   INFO   Env or_v4_5 ready: num_tasks=92 group_scoring=False   ← train env ✅
03:38:07   INFO   Env or_v4_5 ready: num_tasks=92 group_scoring=False   ← eval env ❌ (should be 100)
</code></pre>
<h3>Root Cause (hypothesis)</h3>
<p>Inside <code>src/prime_rl/orchestrator/envs.py</code>, the orchestrator drives the eval environment purely by task index via <code>EnvClient.info()</code>:</p>
<pre><code class="language-python">async def start(self, ...) -&gt; None:
    await super().start(...)
    n = self.num_tasks if self.config.num_examples &lt; 0 else min(self.config.num_examples, self.num_tasks)
    self.examples = [{"task_idx": i} for i in range(n)]
</code></pre>
<p><code>self.num_tasks</code> is populated from the server's <code>info()</code> response. The env server appears to always answer with the size of <code>dataset</code> (the train split), regardless of whether the caller is the train orchestrator or the eval orchestrator. There is currently no mechanism for the eval orchestrator to signal to the env server <em>"use <code>eval_dataset</code> for this session"</em>, so the server never switches its active task pool.</p>
<p>As a result:</p>

  | Expected | Actual
-- | -- | --
TrainEnv.num_tasks | 92 | 92 ✅
EvalEnv.num_tasks | 100 | 92 ❌

<p>This means <strong>eval is silently running against the training split</strong>, which contaminates eval metrics and defeats the purpose of holding out a test set.</p>
<h3>Expected Behaviour</h3>
<p>When the orchestrator spawns (or connects to) an env server for an <code>EvalEnv</code>, the server should expose <code>eval_dataset</code> as its task pool (i.e., <code>info().num_tasks == len(eval_dataset) == 100</code>), and rollouts should be drawn from that split.</p>
<hr>
<blockquote>
<p><strong>Note:</strong> This is a silent failure — no error is raised, logs look healthy, but eval numbers reflect training-data performance. It is easy to miss during a long run. A simple assertion <code>assert eval_env.num_tasks != train_env.num_tasks</code> (when both share the same <code>env_id</code>) would at least surface the issue early.</p>
</blockquote>
</body></html>
